File: demo-typescript/src/shipping-cost-calculator.ts
export class ShippingCostCalculator {
    private static readonly MaxFreeWeightInKg = 5.00;
    private static readonly BaseFee = 1.00;
    private static readonly HeavyWeightFee = 1.50;
    private static readonly ExpressFee = 7.50;

    private cache: Map<string, number> = new Map();

    calculateShippingCost(weightInKg: number, expressShipping: boolean = false): number {
        const cacheKey = `${weightInKg}:${expressShipping}`;

        if (this.cache.has(cacheKey)) {
            return this.cache.get(cacheKey)!;
        }

        let cost = ShippingCostCalculator.BaseFee;

        if (weightInKg > ShippingCostCalculator.MaxFreeWeightInKg) {
            cost += ShippingCostCalculator.HeavyWeightFee;
        }

        if (expressShipping) {
            cost += ShippingCostCalculator.ExpressFee;
        }

        this.cache.set(cacheKey, cost);
        return cost;
    }
}

File: demo-csharp/MutationTesting.Domain/ShippingCostCalculator.cs
namespace MutationTesting.Domain;

public class ShippingCostCalculator
{
    const decimal MaxFreeWeightInKg = 5.00m;

    const decimal BaseFee = 1.00M;
    const decimal HeavyWeightFee = 1.50m;
    const decimal ExpressFee = 7.50m;

    private readonly Dictionary<string, decimal> _cache = new();

    public decimal CalculateShippingCost(decimal weightInKg, bool expressShipping = false)
    {
        var cacheKey = $"{weightInKg}:{expressShipping}";

        if (_cache.TryGetValue(cacheKey, out var cachedCost))
        {
            return cachedCost;
        }

        var cost = BaseFee;

        if (weightInKg > MaxFreeWeightInKg)
        {
            cost += HeavyWeightFee;
        }

        if (expressShipping)
        {
            cost += ExpressFee;
        }

        _cache[cacheKey] = cost;

        return cost;
    }
}